account/seed.go

```diff
@@ -0,0 +1,206 @@
+	expectedChecksum := digest[:seedChecksumLength]
+	actualChecksum := seed[checksumStart:]
+	if !bytes.Equal(expectedChecksum, actualChecksum) {
+		return nil, fault.ErrChecksumMismatch
+	}
+
+	header := seed[:seedHeaderLength]
+	var ed25519Seed []byte // the ed25519 seed for generate key pair
+	var testnet bool       // denote the network is test net
+
+	switch {
+	case bytes.Equal(seedHeaderV1, header):
+		// copy the secret key from seed
+		var sk [secretKeyV1Length]byte
+		secretStart := seedHeaderLength + seedPrefixLength
+		copy(sk[:], seed[secretStart:])
+
+		prefix := seed[seedHeaderLength:secretStart]
+		// first byte of prefix is test/live indication
+		testnet = prefix[0] == 0x01
+
+		ed25519Seed = secretbox.Seal([]byte{}, authSeedIndex[:], &seedNonce, &sk)
+
+	case bytes.Equal(seedHeaderV2, header):
+		sk := seed[seedHeaderLength:checksumStart]
+
+		// verify valid secret key
+		if secretKeyV2Length != len(sk) || 0 != sk[16]&0x0f {
+			return nil, fault.ErrInvalidSeedLength
+		}
+
+		// parse network
+		mode := sk[0]&0x80 | sk[1]&0x40 | sk[2]&0x20 | sk[3]&0x10
+		testnet = mode == sk[15]&0xf0^0xf0
+
+		// add the seed 4 times to hash value
+		hash := sha3.NewShake256()
+		for i := 0; i < 4; i++ {
+			n, err := hash.Write(sk)
+			if err != nil {
+				return nil, err
+			}
+			if secretKeyV2Length != n {
+				return nil, fault.ErrCannotDecodeSeed
+			}
+		}
+
+		ed25519Seed = make([]byte, ed25519.SeedSize)
+		n, err := hash.Read(ed25519Seed)
+		if nil != err {
+			return nil, err
+		}
+		if ed25519.SeedSize != n {
+			return nil, fault.ErrCannotDecodeSeed
+		}
+
+	default:
+		return nil, fault.ErrInvalidSeedHeader
+	}
+
+	// generate key pair from encrypted secret key
+	_, priv, err := ed25519.GenerateKey(bytes.NewBuffer(ed25519Seed))
+	if nil != err {
+		return nil, err
+	}
+
+	privateKey := &PrivateKey{
+		PrivateKeyInterface: &ED25519PrivateKey{
+			Test:       testnet,
+			PrivateKey: priv,
+		},
+	}
+	return privateKey, nil
+}
+
+// NewSeedV1 - generate base58 seed v1
+func NewSeedV1(testnet bool) (string, error) {
+	// generate new seed
+	seedCore := make([]byte, secretKeyV1Length)
+	n, err := rand.Read(seedCore)
+	if nil != err {
+		return "", err
+	}
+	if secretKeyV1Length != n {
+		panic("too few random bytes")
+	}
+	net := 0x00
+	if testnet {
+		net = 0x01
+	}
+	packedSeed := append(seedHeaderV1, byte(net))
+	packedSeed = append(packedSeed, seedCore...)
+	checksum := sha3.Sum256(packedSeed)
+	packedSeed = append(packedSeed, checksum[:seedChecksumLength]...)
+
+	seed := util.ToBase58(packedSeed)
+	return seed, nil
+}
+
+// NewSeedV2 - generate base58 seed v2
+func NewSeedV2(testnet bool) (string, error) {
+
+	// space for 128 bit, extend to 132 bit later
+	sk := make([]byte, 16, secretKeyV2Length)
+
+	n, err := rand.Read(sk)
+	if nil != err {
+		return "", err
+	}
+
+	if 16 != n {
+		return "", fmt.Errorf("got %d bytes, expected is 16 bytes", n)
+	}
+
+	// extend to 132 bits
+	sk = append(sk, sk[15]&0xf0)
```

Comment:
I am not sure the reason to make length to 16 and extend to 17 (secretKeyV2Length), with additional array copy, is it possible to use

```
sk := make([]byte, secretKeyV2Lengt)
n, err := rand.Read(sk)
...
sk[secretKeyV2Length - 1] = sk[secretKeyV2Length - 2]&0xf0
```

Reply:
random read will get 17 bytes (waste entropy) not sure how much extra overhead 1 extra byte of random is.
the real intention is to show that only 128 bits of rando are actually used

Reply:
Maybe better comment needed

Reply:
ok, got it

Reply:
only 16 bytes contain random data, the 17 byte only contains 4 bits derived from the 16 bytes.  I think it is confusing to read 17 bytes then overwrite one